drivers: serial: nrfx_uarte: Add support to DMM  and data cache and other minor tweaks

nordic-krch opened the pull request:
Add support for DMM and cache.
Add configurable magic word used for RX flush workaround.
Add workaround for lack of retention of the BAUDRATE register in UARTE120.
Add lock to uart_rx_disable.

Fixes #77060.

nordic-krch committed:
drivers: serial: nrfx_uarte: Improve RX FIFO flush workaround

Add configurable magic byte instead of fixed 0xAA. In some cases, it
is known that certain bytes are less likely in the transmission and
picking specific magic byte may reduce probability of failure of
detection of the correct amount of flushed data.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add support for DMM and cache

Add support for DMM which manages cache and dedicated memory spaces.
Added support for data cache for buffers which are not DMM managed.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add workaround for BAUDRATE register

On uart120 BAUDRATE register is not retained when ENABLE=0 and
because of that BAUDRATE must be set after enabling. Add workaround
to the driver.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add lock to uart_rx_disable

Add lock to fix race when uart_rx_disable is interrupted by RXTO
event which lead to driver state corruption.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -15,19 +15,28 @@
 #include <zephyr/irq.h>
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(uart_nrfx_uarte, CONFIG_UART_LOG_LEVEL);
+#ifdef CONFIG_HAS_NORDIC_DMM
```
is this really needed? for legacy platforms these macros in question should evaluate to nothing/`0` : 
- https://github.com/zephyrproject-rtos/zephyr/blob/98e0e65069cf62297c4d03e0ad5b543374f5b735/soc/nordic/common/dmm.h#L61
- https://github.com/zephyrproject-rtos/zephyr/blob/98e0e65069cf62297c4d03e0ad5b543374f5b735/soc/nordic/common/dmm.h#L28

nordic-krch replied:
removed.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2227,11 +2376,15 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 				UARTE_CFG_FLAG_PPI_ENDTX : 0) |		       \
 			(IS_ENABLED(CONFIG_UART_##idx##_NRF_HW_ASYNC) ?        \
 				UARTE_CFG_FLAG_HW_BYTE_COUNTING : 0) |	       \
+			(!IS_ENABLED(CONFIG_HAS_NORDIC_DMM) ? 0 :	       \
+			  (UARTE_IS_CACHEABLE(idx) ? UARTE_CFG_FLAG_CACHEABLE : 0)) | \
```
`UARTE_IS_CACHEABLE(idx)` evaluates to `false` when `!CONFIG_HAS_NORDIC_DMM`, so:
```suggestion
			(UARTE_IS_CACHEABLE(idx) ?			       \
				UARTE_CFG_FLAG_CACHEABLE : 0) |		       \
```

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -571,6 +621,10 @@ static void tx_start(const struct device *dev, const uint8_t *buf, size_t len)
 #endif
 
+	if (IS_ENABLED(UARTE_ANY_CACHE) && (config->flags & UARTE_CFG_FLAG_CACHEABLE)) {
+		sys_cache_data_flush_range((void *)buf, len);
```
theoretically you could indiscriminately use `dmm_buffer_out_prepare` that would flush the cache for you and be transparent for the legacy targets. But I guess this could be purposefully omitted for performance reasons

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -877,6 +931,19 @@ static int uarte_nrfx_rx_enable(const struct device *dev, uint8_t *buf,
 		return -EBUSY;
 	}
 
+#ifdef CONFIG_HAS_NORDIC_DMM
```
this could should compile fine for legacy targets as well - `dmm_buffer_in_prepare` is mocked for them: https://github.com/zephyrproject-rtos/zephyr/blob/98e0e65069cf62297c4d03e0ad5b543374f5b735/soc/nordic/common/dmm.h#L134

nordic-krch replied:
im ifdefing because for dmm case there are additional fields in the data and config structure to hold user buffer (as it is different one than one used for DMA) and `mem_reg`. We could keep this for non-dmm cases and that would work but then its redundant so i preferred to ifdef it.

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -909,8 +976,18 @@ static int uarte_nrfx_rx_enable(const struct device *dev, uint8_t *buf,
 
+			if (IS_ENABLED(UARTE_ANY_CACHE) &&
+			    (cfg->flags & UARTE_CFG_FLAG_CACHEABLE)) {
+				sys_cache_data_invd_range(cfg->rx_flush_buf, cpy_len);
```
I guess this shouldn't be covered by `dmm`, as this is a case in which both `rx_flush_buf` and `buf` are located in cacheable RAM and in fact we are "emulating" DMA copying data from the former to the latter ?

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -921,6 +998,11 @@ static int uarte_nrfx_rx_enable(const struct device *dev, uint8_t *buf,
 						async_rx->flush_cnt);
+				if (IS_ENABLED(UARTE_ANY_CACHE) &&
+				    (cfg->flags & UARTE_CFG_FLAG_CACHEABLE)) {
+					sys_cache_data_flush_range(cfg->rx_flush_buf,
```
as above - `dmm` shouldn't be used here as this is a shim-specific code maintaining `rx_flush_buf` that happens to be located in cached RAM, right ?

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -966,6 +1048,17 @@ static int uarte_nrfx_rx_buf_rsp(const struct device *dev, uint8_t *buf,
 	if (async_rx->buf == NULL) {
 		err = -EACCES;
 	} else if (async_rx->next_buf == NULL) {
+#ifdef CONFIG_HAS_NORDIC_DMM
```
as above - this could should compile fine for legacy targets as well - `dmm_buffer_in_prepare` is mocked for them: https://github.com/zephyrproject-rtos/zephyr/blob/98e0e65069cf62297c4d03e0ad5b543374f5b735/soc/nordic/common/dmm.h#L134

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1235,6 +1334,15 @@ static void endrx_isr(const struct device *dev)
 	 */
 	const int rx_amount = nrf_uarte_rx_amount_get(uarte) + async_rx->flush_cnt;
 
+#ifdef CONFIG_HAS_NORDIC_DMM
```
as above - this could should compile fine for legacy targets as well - `dmm_buffer_in_release` is mocked for them

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -112,6 +137,10 @@ struct uarte_async_tx {
 
 struct uarte_async_rx {
 	struct k_timer timer;
+#ifdef CONFIG_HAS_NORDIC_DMM
```
there are several comments regarding indiscriminate usage of `dmm` APIs everywhere VS small potential savings in state variables. If you find them valuable they can stay, I guess

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -57,6 +57,16 @@ config UART_ASYNC_TX_CACHE_SIZE
 
+config UART_NRFX_UARTE_RX_FLUSH_MAGIC_BYTE
+	int "Byte used for RX FIFO flush workaround"
+	default 171
```
Was it intentional that now the default magic byte is different (0xAB) than previously used 0xAA?
And wouldn't a hexadecimal number be more natural here?

nordic-krch replied:
Yes, it was intentional, as BT guys asked to use something different than 0xAA because their tests were hitting an error due to 0xAA use. However, today I've got information from them that 0xAB is also not good. We will be looking for another solution to their issue so I get back here to 0xAA. I was not aware of `hex`. Changed to use that.

nika-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1633,6 +1766,10 @@ static int uarte_nrfx_poll_in(const struct device *dev, unsigned char *c)
 	}
 
+	if (IS_ENABLED(UARTE_ANY_CACHE) && (config->flags & UARTE_CFG_FLAG_CACHEABLE)) {
+		sys_cache_data_invd_range(config->poll_in_byte, 1);
```
`poll_in_byte` could be managed by `dmm` to avoid manual cache management in this case. `dmm_buffer_in_prepare` would need to be called before starting the receiver and `dm_buffer_in_release` on ENDRX. But again, this looks like special case as this single byte is used internally by shim while `dmm` was rather designed for user-provided buffers

nika-nordic reviewed (dismissed):
Left some comments regarding `dmm` usage. It could be debated how widely it should be used in this shim, however rule of thumb is that `dmm` was designed for user-provided buffers, while UARTE shim has some internal buffers ( `rx_flush_buf`, `poll_in_byte` ) where `dmm` could be an overkill

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2042,12 +2183,19 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 
 		nrf_uarte_enable(uarte);
+#if UARTE_BAUDRATE_RETENTION_WORKAROUND
+		nrf_uarte_baudrate_t baudrate = COND_CODE_1(CONFIG_UART_USE_RUNTIME_CONFIGURE,
+			(data->nrf_baudrate), (cfg->nrf_baudrate));
+
+		nrf_uarte_baudrate_set(get_uarte_instance(dev), baudrate);
```
Why not just:
```suggestion
		nrf_uarte_baudrate_set(get_uarte_instance(dev),
			COND_CODE_1(CONFIG_UART_USE_RUNTIME_CONFIGURE,
				(data->nrf_baudrate), (cfg->nrf_baudrate)));
```
to avoid those extra braces?

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -57,6 +57,16 @@ config UART_ASYNC_TX_CACHE_SIZE
 
+config UART_NRFX_UARTE_RX_FLUSH_MAGIC_BYTE
+	int "Byte used for RX FIFO flush workaround"
+	default 170 # 0xAA
```
`default 0xAA` should also work.

nordic-krch commented:
@anangl your comments applied.
@nika-nordic some comments applied. Due to resources usage I would stay with what we got regarding use of dmm where it is not needed.

nordic-krch committed:
boards: native: nrf_bsim: Add include path to common

Add include path to nordic/soc/nordic/common where some internal
headers for nordic devices are kept (e.g. dmm.h).

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
After including dmm.h unconditionally `nrf52_bsim` target compilation started to fail due to lack of `soc/nordic/common` include path. I've added it to bsim boards. @aescolar can you take a look if that is ok?

aescolar commented:
> I've added it to bsim boards. @aescolar can you take a look if that is ok?

@nordic-krch I think that is fine
